pilot/pkg/proxy/envoy/v2/grpc.go

```diff
@@ -0,0 +1,262 @@
+		TypeUrl:     ClusterType,
+		VersionInfo: version,
+		Nonce:       nonce(push.Version),
+	}
+
+	// gRPC doesn't currently support any of the APIs - returning just the expected EDS result.
+	// Since the code is relatively strict - we'll add info as needed.
+	for _, n := range req.ResourceNames {
+		hn, portn, err := net.SplitHostPort(n)
+		if err != nil {
+			adsLog.Warna("Failed to parse ", n, " ", err)
+			continue
+		}
+		rc := &xdsapi.Cluster{
+			Name: n,
+			ClusterDiscoveryType: &xdsapi.Cluster_Type{Type: xdsapi.Cluster_EDS},
+			EdsClusterConfig: &xdsapi.Cluster_EdsClusterConfig{
+				ServiceName: "outbound|" + portn + "||" + hn,
+				EdsConfig: &envoycore.ConfigSource{
+					ConfigSourceSpecifier: &envoycore.ConfigSource_Ads{
+						Ads: &envoycore.AggregatedConfigSource{},
+					},
+
+				},
+			},
+		}
+		rr := util.MessageToAny(rc)
+		resp.Resources = append(resp.Resources, rr)
+	}
+
+	err := con.send(resp)
+	if err != nil {
+		adsLog.Warnf("LDS: Send failure %s: %v", con.ConID, err)
+		recordSendError(ldsSendErrPushes, err)
+		return true
+	}
+	return true
+}
+
+// handleSplitRDS supports per-VIP routes, as used by GRPC.
+// This mode is indicated by using names containing full host:port instead of just port.
+// Returns true of the request is of this type.
+func (s *DiscoveryServer) handleSplitRDS(con *XdsConnection, req *xdsapi.DiscoveryRequest) bool {
+	for _, n := range req.ResourceNames {
+		if !strings.Contains(n, ":") {
+			return false // normal Istio RDS, on port
+		}
+	}
+
+	if s.handleAck(con, req) {
+		return true
+	}
+	push := s.globalPushContext()
+
+	// Currently this mode is only used by GRPC, to extract Cluster for the default
+	// route.
+	// Current GRPC is also expecting the default route to be prefix=="", while we generate "/"
+	// in normal response.
+	// TODO: add support for full route, make sure GRPC is fixed to support both
+	resp := &xdsapi.DiscoveryResponse{
+		TypeUrl:     RouteType,
+		VersionInfo: version,
+		Nonce:       nonce(push.Version),
+	}
+	for _, n := range req.ResourceNames {
+		hn, portn, err := net.SplitHostPort(n)
+		if err != nil {
+			adsLog.Warna("Failed to parse ", n, " ", err)
+			continue
+		}
+		port, err := strconv.Atoi(portn)
+		if err != nil {
+			adsLog.Warna("Failed to parse port ", n, " ", err)
+			continue
+		}
+		el := con.node.SidecarScope.GetEgressListenerForRDS(port, "")
+		vsvc := el.VirtualServices()
+		for _, s:= range vsvc {
+			vs, ok := s.Spec.(*networking.VirtualService)
+			if !ok {
+				continue
+			}
+			for _, vsh := range vs.Hosts {
+				if vsh == hn {
+					// Only generate the required route for grpc. Will need to generate more
+					// as GRPC adds more features.
+					rc := &envoy_api_v2_route.Route{
+						Match: &envoy_api_v2_route.RouteMatch{
+							PathSpecifier:        &envoy_api_v2_route.RouteMatch_Prefix{Prefix:""},
+						},
+					}
+					rr := util.MessageToAny(rc)
+					resp.Resources = append(resp.Resources, rr)
```

Comment:
How are we adding `Route` here and `RouteConfiguration` below? Don't we expect only one type in response?

Reply:
Right - I was trying to implement VirtualService part - cleaned up and added a comment. 

It should be RouteConfiguration.